From 2c48917c4090eca6c6aed12acdee940d0f2e757b Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Thu, 18 Jul 2024 19:31:11 +0200
Subject: [PATCH] Add CountTxSigOps (+P2SH and NonP2SH variants) (#23)

These count the sig ops of the given transaction (using the old sigops rules used in Dogecoin).

(cherry picked from commit 1873d75695ab2fd2b1141c347699d85651a1efbe)

diff --git a/src/script/sigops.cpp b/src/script/sigops.cpp
index bcfa0a829800622c26d3c0426dc4d3b3fd39a2cf..9fb4fc2798201b2c182a802edbe0a8c784ee1fca 100644
--- a/src/script/sigops.cpp
+++ b/src/script/sigops.cpp
@@ -2,6 +2,8 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+#include <coins.h>
+#include <primitives/transaction.h>
 #include <script/script.h>
 #include <script/sigops.h>
 
@@ -68,3 +70,36 @@ uint32_t CountScriptSigOpsP2SH(const CScript &scriptSig) {
     CScript subscript(vData.begin(), vData.end());
     return CountScriptSigOps(subscript, SigOpCountMode::ACCURATE);
 }
+
+uint64_t CountTxNonP2SHSigOps(const CTransaction &tx) {
+    uint64_t nSigOps = 0;
+    for (const CTxIn &txin : tx.vin) {
+        nSigOps += CountScriptSigOps(txin.scriptSig, SigOpCountMode::ESTIMATED);
+    }
+    for (const CTxOut &txout : tx.vout) {
+        nSigOps +=
+            CountScriptSigOps(txout.scriptPubKey, SigOpCountMode::ESTIMATED);
+    }
+    return nSigOps;
+}
+
+uint64_t CountTxP2SHSigOps(const CTransaction &tx,
+                           const CCoinsViewCache &view) {
+    if (tx.IsCoinBase()) {
+        return 0;
+    }
+
+    uint64_t nSigOps = 0;
+    for (const CTxIn &txin : tx.vin) {
+        const Coin &coin = view.AccessCoin(txin.prevout);
+        if (coin.GetTxOut().scriptPubKey.IsPayToScriptHash()) {
+            nSigOps += CountScriptSigOpsP2SH(txin.scriptSig);
+        }
+    }
+
+    return nSigOps;
+}
+
+uint64_t CountTxSigOps(const CTransaction &tx, const CCoinsViewCache &view) {
+    return CountTxNonP2SHSigOps(tx) + CountTxP2SHSigOps(tx, view);
+}
diff --git a/src/script/sigops.h b/src/script/sigops.h
index 326008406708ae169afe362383efdc9e4edda72a..9930cdb03d557076d3d0e08e38e8eb1d29bb06e2 100644
--- a/src/script/sigops.h
+++ b/src/script/sigops.h
@@ -3,6 +3,8 @@
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
 class CScript;
+class CTransaction;
+class CCoinsViewCache;
 
 enum class SigOpCountMode {
     ACCURATE,
@@ -12,3 +14,9 @@ enum class SigOpCountMode {
 uint32_t CountScriptSigOps(const CScript &script, SigOpCountMode mode);
 
 uint32_t CountScriptSigOpsP2SH(const CScript &scriptSig);
+
+uint64_t CountTxNonP2SHSigOps(const CTransaction &tx);
+
+uint64_t CountTxP2SHSigOps(const CTransaction &tx, const CCoinsViewCache &view);
+
+uint64_t CountTxSigOps(const CTransaction &tx, const CCoinsViewCache &view);
diff --git a/src/test/sigops_tests.cpp b/src/test/sigops_tests.cpp
index 892195d14d95b4ab9705fb030ad7613499eea4aa..9aeddfff1cd167ad09195456441a222524ff4434 100644
--- a/src/test/sigops_tests.cpp
+++ b/src/test/sigops_tests.cpp
@@ -4,6 +4,7 @@
 
 #include <boost/test/unit_test.hpp>
 
+#include <coins.h>
 #include <script/script.h>
 #include <script/sigops.h>
 #include <script/standard.h>
@@ -116,4 +117,120 @@ BOOST_AUTO_TEST_CASE(CountScriptSigOpsP2SH_test) {
     }
 }
 
+/**
+ * Builds a creationTx from scriptPubKey and a spendingTx from scriptSig
+ * such that spendingTx spends output zero of creationTx. Also inserts
+ * creationTx's output into the coins view.
+ */
+static void BuildTxs(CMutableTransaction &spendingTx, CCoinsViewCache &coins,
+                     CMutableTransaction &creationTx,
+                     const CScript &scriptPubKey, const CScript &scriptSig) {
+    creationTx.nVersion = 1;
+    creationTx.vin.resize(1);
+    creationTx.vin[0].prevout = COutPoint();
+    creationTx.vin[0].scriptSig = CScript();
+    creationTx.vout.resize(1);
+    creationTx.vout[0].nValue = SATOSHI;
+    creationTx.vout[0].scriptPubKey = scriptPubKey;
+
+    spendingTx.nVersion = 1;
+    spendingTx.vin.resize(1);
+    spendingTx.vin[0].prevout = COutPoint(creationTx.GetId(), 0);
+    spendingTx.vin[0].scriptSig = scriptSig;
+    spendingTx.vout.resize(1);
+    spendingTx.vout[0].nValue = SATOSHI;
+    spendingTx.vout[0].scriptPubKey = CScript();
+
+    AddCoins(coins, CTransaction(creationTx), 0);
+}
+
+BOOST_AUTO_TEST_CASE(CountTxSigOps_test) {
+    // Transaction creates outputs
+    CMutableTransaction creationTx;
+    // Transaction that spends outputs and whose sig op count is going to be
+    // tested
+    CMutableTransaction spendingTx;
+
+    std::vector<uint8_t> dummy(20);
+
+    // Coinbase
+    {
+        CCoinsView coinsDummy;
+        CCoinsViewCache coins(&coinsDummy);
+
+        // Coinbase with a OP_CHECKSIG in the sig script is counted
+        creationTx.nVersion = 1;
+        creationTx.vin.resize(1);
+        creationTx.vin[0].prevout = COutPoint();
+        creationTx.vin[0].scriptSig = CScript() << OP_CHECKSIG;
+        CTransaction createTx(creationTx);
+        BOOST_CHECK_EQUAL(CountTxSigOps(createTx, coins), 1);
+        BOOST_CHECK_EQUAL(CountTxNonP2SHSigOps(createTx), 1);
+        BOOST_CHECK_EQUAL(CountTxP2SHSigOps(createTx, coins), 0);
+
+        // Coinbase with something looking like a P2SH redeemScript not counted
+        creationTx.vin[0].scriptSig = CScript()
+                                      << ToByteVector(CScript() << OP_CHECKSIG);
+        CTransaction createTx2(creationTx);
+        BOOST_CHECK_EQUAL(CountTxSigOps(createTx2, coins), 0);
+        BOOST_CHECK_EQUAL(CountTxNonP2SHSigOps(createTx2), 0);
+        BOOST_CHECK_EQUAL(CountTxP2SHSigOps(createTx2, coins), 0);
+    }
+
+    // Multisig script (legacy counting)
+    {
+        CCoinsView coinsDummy;
+        CCoinsViewCache coins(&coinsDummy);
+
+        CScript scriptPubKey = CScript() << 1 << dummy << dummy << 2
+                                         << OP_CHECKMULTISIGVERIFY;
+        // Do not use a valid signature to avoid using wallet operations.
+        CScript scriptSig = CScript() << OP_0 << OP_0;
+
+        BuildTxs(spendingTx, coins, creationTx, scriptPubKey, scriptSig);
+        CTransaction createTx(creationTx);
+        CTransaction spendTx(spendingTx);
+
+        // Legacy counting only includes signature operations in scriptSigs and
+        // scriptPubKeys of a transaction and does not take the actual executed
+        // sig operations into account. spendingTx in itself does not contain a
+        // signature operation.
+        BOOST_CHECK_EQUAL(CountTxSigOps(spendTx, coins), 0);
+        BOOST_CHECK_EQUAL(CountTxP2SHSigOps(spendTx, coins), 0);
+        BOOST_CHECK_EQUAL(CountTxNonP2SHSigOps(spendTx), 0);
+        // creationTx contains two signature operations in its scriptPubKey, but
+        // legacy counting is not accurate.
+        BOOST_CHECK_EQUAL(CountTxSigOps(createTx, coins),
+                          MAX_PUBKEYS_PER_MULTISIG);
+        BOOST_CHECK_EQUAL(CountTxP2SHSigOps(createTx, coins), 0);
+        BOOST_CHECK_EQUAL(CountTxNonP2SHSigOps(createTx),
+                          MAX_PUBKEYS_PER_MULTISIG);
+    }
+
+    // Multisig nested in P2SH
+    {
+        CCoinsView coinsDummy;
+        CCoinsViewCache coins(&coinsDummy);
+
+        CScript redeemScript = CScript() << 1 << dummy << dummy << 2
+                                         << OP_CHECKMULTISIGVERIFY;
+        CScript scriptPubKey =
+            GetScriptForDestination(ScriptHash(redeemScript));
+        CScript scriptSig = CScript()
+                            << OP_0 << OP_0 << ToByteVector(redeemScript);
+
+        BuildTxs(spendingTx, coins, creationTx, scriptPubKey, scriptSig);
+        CTransaction createTx(creationTx);
+        CTransaction spendTx(spendingTx);
+
+        BOOST_CHECK_EQUAL(CountTxSigOps(spendTx, coins), 2);
+        BOOST_CHECK_EQUAL(CountTxP2SHSigOps(spendTx, coins), 2);
+        BOOST_CHECK_EQUAL(CountTxNonP2SHSigOps(spendTx), 0);
+
+        BOOST_CHECK_EQUAL(CountTxSigOps(createTx, coins), 0);
+        BOOST_CHECK_EQUAL(CountTxP2SHSigOps(createTx, coins), 0);
+        BOOST_CHECK_EQUAL(CountTxNonP2SHSigOps(createTx), 0);
+    }
+}
+
 BOOST_AUTO_TEST_SUITE_END()
